gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -106,15 +119,60 @@ def self.signature_for_method(method)
     @signatures_by_method[key]
   end
 
+  # when `mod` includes a module with instance methods `method_names`, ensure there is zero intersection between the
+  # final instance methods of `mod` and `method_names`. so, for every m in `method_names`, check if there is already a
+  # method defined on one of `mod`'s `ancestors` with the same name that is final.
+  def self._check_final_ancestors(mod, ancestors, method_names)
+    # use reverse_each to check farther-up ancestors first, for better error messages. we could avoid this if we were on
+    # the version of ruby that adds the optional argument to method_defined? that allows you to exclude ancestors.
+    ancestors.reverse_each do |ancestor|
+      method_names.each do |method_name|
+        # the usage of method_owner_and_name_to_key(ancestor, method_name) instead of
+        # method_to_key(ancestor.instance_method(method_name)) is not (just) an optimization, but also required for
+        # correctness, since ancestor.method_defined?(method_name) may return true even if method_name is not defined
+        # directly on ancestor but instead an ancestor of ancestor.
+        if ancestor.method_defined?(method_name) && final_method?(method_owner_and_name_to_key(ancestor, method_name))
+          raise(
+            "`#{ancestor.name}##{method_name}` was declared as final and cannot be " +
+            (mod == ancestor ? "redefined" : "overridden in `#{mod.name}`")
+          )
+        end
+      end
+    end
+  end
+
+  private_class_method def self.add_final_method(method_key)
+    @final_methods.add(method_key)
+  end
+
+  private_class_method def self.final_method?(method_key)
+    @final_methods.include?(method_key)
+  end
+
+  def self.add_module_with_final(mod)
```

Comment:
That's because I have the final module branch on top of this branch, and in the final module branch, I want to mark modules declared as final with `final!` as being modules containing final things.